src/System.Management.Automation/engine/runtime/Profiler.cs

```diff
@@ -0,0 +1,402 @@
+                        Size = sizeof(Guid),
+                        DataPointer = ((IntPtr)(&ScriptBlockId))
+                    };
+                    eventPayload[1] = new EventData
+                    {
+                        Size = sizeof(int),
+                        DataPointer = ((IntPtr)(&SequencePointPosition))
+                    };
+
+                    WriteEventCore(eventId: 1, eventDataCount: 2, eventPayload);
+                }
+            }
+        }
+
+        protected override void OnEventCommand(EventCommandEventArgs command)
+        {
+            base.OnEventCommand(command);
+
+            if (command.Command == EventCommand.Disable)
+            {
+                // At the end of the profile session, we send all the metadata.
+                ProfilerRundownEventSource.LogInstance.WriteRundownEvents();
+            }
+        }
+    }
+
+    /// <summary>
+    /// This class is a source of rundown (meta data) events for script block profiling.
+    /// Guid is {f348266e-dde6-5590-4bc9-10e1e2b6fe16}.
+    /// </summary>
+    [EventSource(Name = "Microsoft-PowerShell-Profiler-Rundown")]
+    internal class ProfilerRundownEventSource : EventSource
+    {
+        internal static ProfilerRundownEventSource LogInstance = new ProfilerRundownEventSource();
+
+        [Event(2)]
+        public void ScriptBlockRundown(
+            Guid ScriptBlockId,
+            string ScriptBlockText)
+        {
+            // It is not performance critical so we use the standard method overload.
+            WriteEvent(
+                eventId: 2,
+                ScriptBlockId,
+                ScriptBlockText);
+        }
+
+        [Event(3)]
+        public void SequencePointRundown(
+            Guid ScriptBlockId,
+            int SequencePointCount,
+            int SequencePoint,
+            string File,
+            int StartLineNumber,
+            int StartColumnNumber,
+            int EndLineNumber,
+            int EndColumnNumber,
+            string Text,
+            int StartOffset,
+            int EndOffset)
+        {
+            // It is not performance critical so we use the standard method overload.
+            WriteEvent(
+                eventId: 3,
+                ScriptBlockId,
+                SequencePointCount,
+                SequencePoint,
+                File,
+                StartLineNumber,
+                StartColumnNumber,
+                EndLineNumber,
+                EndColumnNumber,
+                Text,
+                StartOffset,
+                EndOffset);
+        }
+
+        [NonEvent]
+        internal void WriteRundownEvents()
+        {
+            foreach (var pair in CompiledScriptBlockData.GetCompiledScriptBlockTable())
+            {
+                var compiledScriptBlock = pair.Key;
+                ScriptBlockRundown(compiledScriptBlock.Id, compiledScriptBlock.Ast.Body.Extent.Text);
+
+                if (compiledScriptBlock.SequencePoints is null)
```

Comment:
Based on my comment and the code it guards, I think the point was to avoid needing tests like this elsewhere.

Reply:
But I catch NRE here :-( Should we consider this as separate bug and open a tracking issue?

Reply:
I suppose it's not exactly a bug if it's not causing other issues now - so your call.